File: db4oj.tests/src/com/db4o/db4ounit/common/concurrency/MultiDeleteTestCase.java
/* Copyright (C) 2004 - 2007  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.concurrency;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class MultiDeleteTestCase extends Db4oClientServerTestCase {

	public static void main(String[] args) {
		new MultiDeleteTestCase().runConcurrency();
	}
	
	public MultiDeleteTestCase child;

	public String name;

	public Object forLong;

	public Long myLong;

	public Object[] untypedArr;

	public Long[] typedArr;

	protected void configure(Configuration config) {
		config.objectClass(this).cascadeOnDelete(true);
		config.objectClass(this).cascadeOnUpdate(true);
	}

	protected void store() {
		MultiDeleteTestCase md = new MultiDeleteTestCase();
		md.name = "killmefirst";
		md.setMembers();
		md.child = new MultiDeleteTestCase();
		md.child.setMembers();
		store(md);
	}

	public void conc(ExtObjectContainer oc) throws Exception {
		Query q = oc.query();
		q.constrain(MultiDeleteTestCase.class);
		q.descend("name").constrain("killmefirst");
		ObjectSet objectSet = q.execute();
		if (objectSet.size() == 0) { // already deleted by other threads
			return;
		}
		
		Assert.areEqual(1, objectSet.size());
		Thread.sleep(1000);
		if(!objectSet.hasNext()){
			return;
		}
		
		MultiDeleteTestCase md = (MultiDeleteTestCase) objectSet.next();
		oc.delete(md);
		oc.commit();
		assertOccurrences(oc, MultiDeleteTestCase.class, 0);
	}

	public void check(ExtObjectContainer oc) {
		assertOccurrences(oc, MultiDeleteTestCase.class, 0);
	}

	private void setMembers() {
		forLong = new Long(100);
		myLong = new Long(100);
		untypedArr = new Object[] { new Long(10), "hi", new MultiDeleteTestCase() };
		typedArr = new Long[] { new Long(3), new Long(7), new Long(9), };
	}

}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Concurrency/MultiDeleteTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

#if !SILVERLIGHT
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Concurrency;
using Sharpen.Lang;

namespace Db4objects.Db4o.Tests.Common.Concurrency
{
	public class MultiDeleteTestCase : Db4oClientServerTestCase
	{
		public static void Main(string[] args)
		{
			new MultiDeleteTestCase().RunConcurrency();
		}

		public MultiDeleteTestCase child;

		public string name;

		public object forLong;

		public long myLong;

		public object[] untypedArr;

		public long[] typedArr;

		protected override void Configure(IConfiguration config)
		{
			config.ObjectClass(this).CascadeOnDelete(true);
			config.ObjectClass(this).CascadeOnUpdate(true);
		}

		protected override void Store()
		{
			MultiDeleteTestCase md = new MultiDeleteTestCase();
			md.name = "killmefirst";
			md.SetMembers();
			md.child = new MultiDeleteTestCase();
			md.child.SetMembers();
			Store(md);
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void Conc(IExtObjectContainer oc)
		{
			IQuery q = oc.Query();
			q.Constrain(typeof(MultiDeleteTestCase));
			q.Descend("name").Constrain("killmefirst");
			IObjectSet objectSet = q.Execute();
			if (objectSet.Count == 0)
			{
				// already deleted by other threads
				return;
			}
			Assert.AreEqual(1, objectSet.Count);
			Thread.Sleep(1000);
			if (!objectSet.HasNext())
			{
				return;
			}
			MultiDeleteTestCase md = (MultiDeleteTestCase)objectSet.Next();
			oc.Delete(md);
			oc.Commit();
			AssertOccurrences(oc, typeof(MultiDeleteTestCase), 0);
		}

		public virtual void Check(IExtObjectContainer oc)
		{
			AssertOccurrences(oc, typeof(MultiDeleteTestCase), 0);
		}

		private void SetMembers()
		{
			forLong = System.Convert.ToInt64(100);
			myLong = System.Convert.ToInt64(100);
			untypedArr = new object[] { System.Convert.ToInt64(10), "hi", new MultiDeleteTestCase
				() };
			typedArr = new long[] { System.Convert.ToInt64(3), System.Convert.ToInt64(7), System.Convert.ToInt64
				(9) };
		}
	}
}
#endif // !SILVERLIGHT
